instrumentation/mongo/mongo-common/src/main/java/io/opentelemetry/javaagent/instrumentation/mongo/MongoClientTracer.java

```diff
@@ -80,54 +97,111 @@ protected String dbConnectionString(CommandStartedEvent event) {
+  private JsonWriterSettings createJsonWriterSettings(int maxNormalizedQueryLength)  {
+    JsonWriterSettings settings = new JsonWriterSettings(false);
+    try {
+      // The static JsonWriterSettings.builder() method was introduced in the 3.5 release
+      Optional<Method> buildMethod = Arrays.stream(JsonWriterSettings.class.getMethods())
+          .filter(method -> method.getName().equals("builder")).findFirst();
+      if (buildMethod.isPresent()) {
+        Class<?> builderClass = buildMethod.get().getReturnType();
+        Object builder = buildMethod.get().invoke(null, (Object[]) null);
+
+        // The JsonWriterSettings.Builder.indent method was introduced in the 3.5 release,
+        // but checking anyway
+        Optional<Method> indentMethod = Arrays.stream(builderClass.getMethods())
+            .filter(method -> method.getName().equals("indent"))
+            .findFirst();
+        if (indentMethod.isPresent()) {
+          indentMethod.get().invoke(builder, false);
+        }
 
-  private static final BsonValue HIDDEN_CHAR = new BsonString("?");
+        // The JsonWriterSettings.Builder.maxLength method was introduced in the 3.7 release
+        Optional<Method> maxLengthMethod = Arrays.stream(builderClass.getMethods())
+            .filter(method -> method.getName().equals("maxLength"))
+            .findFirst();
+        if (maxLengthMethod.isPresent()) {
+          maxLengthMethod.get().invoke(builder, maxNormalizedQueryLength);
+        }
+        settings = (JsonWriterSettings) builderClass.getMethod("build", (Class<?>[]) null)
+            .invoke(builder, (Object[]) null);
+      }
+    } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException ignored) {
+    }
+    return settings;
+  }
 
-  private static BsonDocument scrub(BsonDocument origin) {
-    BsonDocument scrub = new BsonDocument();
+  @Override
+  public String normalizeQuery(BsonDocument command) {
+    StringWriter stringWriter = new StringWriter(128);
+    writeScrubbed(command, new JsonWriter(stringWriter, jsonWriterSettings), true);
+    // If using MongoDB driver >= 3.7, the substring invocation will be a no-op due to use of
+    // JsonWriterSettings.Builder.maxLength in the static initializer for JSON_WRITER_SETTINGS
+    return stringWriter.getBuffer()
+        .substring(0, Math.min(maxNormalizedQueryLength, stringWriter.getBuffer().length()));
+  }
+
+  private static final String HIDDEN_CHAR = "?";
+
+  private static boolean writeScrubbed(BsonDocument origin, JsonWriter writer, boolean isRoot) {
+    writer.writeStartDocument();
+    boolean firstField = true;
     for (Map.Entry<String, BsonValue> entry : origin.entrySet()) {
-      if (UNSCRUBBED_FIELDS.contains(entry.getKey()) && entry.getValue().isString()) {
-        scrub.put(entry.getKey(), entry.getValue());
+      writer.writeName(entry.getKey());
+      // the first field of the root document is the command name, so we preserve its value
```

Comment:
Question for reviewers: since we already extract the collection name into the Span, is it even necessary to preserve it in the scrubbed command?

Reply:
preserving it seems good, e.g. table names are preserved when scrubbing sql queries